### DISCUSSION TITLE
creating-a-custom-iconfigurationprovider-in-asp-net-core-to-parse-yaml/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#45] [Announcements]
# creating-a-custom-iconfigurationprovider-in-asp-net-core-to-parse-yaml/

ASP.NET Core ships with a new configuration system. In this post I show how to create a custom configuration provider to load YAML files.

https://andrewlock.net/creating-a-custom-iconfigurationprovider-in-asp-net-core-to-parse-yaml/

**Comment by jeffreywstevens** [comment]
Great article. I noticed you are missing an uppercase T unless that was deliberate. 

The location of the the heading for "he FileParser" is what I noticed.

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at April 27 2018, 01:16, in reply to [Shah Keyur B](https://disqus.com/by/keyur786shah/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/45#discussioncomment-5795098)</em>

---

<p>I suggest you take a different approach - use two "layers" of configuration builders. This example from the docs does pretty much exactly what you want, using EF Core: <a href="https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-2.1&amp;tabs=basicconfiguration#create-an-entity-framework-custom-provider" rel="nofollow noopener" title="https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-2.1&amp;tabs=basicconfiguration#create-an-entity-framework-custom-provider">https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-2.1&amp;tabs=basicconfiguration#create-an-entity-framework-custom-provider</a></p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at June 19 2019, 11:44, in reply to [Marco Antonino](https://disqus.com/by/disqus_i1vOiYd4it/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/45#discussioncomment-5795101)</em>

---

<p>Thanks, I really appreciate it! Glad you've found them useful :)</p><p>RE the HashiCorp Vault provider - you're very welcome to take that and run with it :) If you just need a consul configuration provider, I suggest checking out this library though: <a href="https://www.nuget.org/packages/Winton.Extensions.Configuration.Consul/" rel="nofollow noopener" title="https://www.nuget.org/packages/Winton.Extensions.Configuration.Consul/">https://www.nuget.org/packages/Winton.Extensions.Configuration.Consul/</a> has a lot of downloads and looks like it should do the job!</p>

[a comment by an automated account is not shown]

**Comment by dodexahedron** [comment]
One of Andrew's many great series. 👍 

Here's a fun fact: YAML is a superset of JSON, originally by accident/coincidence, but then on purpose, which was formalized 15 years ago in YAML 1.2.

Here's a quick excerpt from the spec that explains when, why, and how it happened:

> The YAML 1.2 specification was published in 2009. Its primary focus was making YAML a strict superset of JSON. It also removed many of the problematic implicit typing recommendations.

And there are quite a few other places in the spec where things are done explicitly to ensure compatibility with JSON.

The upshot is that a valid JSON document is also a valid YAML document, so you should only need one parser to handle both JSON and YAML.

I learned this fairly recently and I remembered this blog post because both concepts are relevant to a current project.

Figured I'd share. :)

**Comment by dodexahedron** [comment]
BTW: The JsonConfigurationFileParser link is no longer valid.
